### ISSUE TITLE
Merge composer.json of plugins and themes

### ISSUE THREAD

**Issue by z38** [getgrav/grav#285]
As mentioned in #32, Grav currently does not support Composer dependencies of plugins and themes. In the meantime, [MediaWiki](https://www.mediawiki.org) also had to deal with almost the same problem. They created [wikimedia/composer-merge-plugin](https://github.com/wikimedia/composer-merge-plugin) which allows multiple `composer.json` files to be merged together.

By extending the current `composer.json`, this feature could be enabled rather easily:

```
{
    "require": {
        "wikimedia/composer-merge-plugin": "dev-master"
    },
    "extra": {
        "merge-plugin": {
            "include": [
                "user/themes/*/composer.json",
                "user/plugins/*/composer.json"
            ],
            "recurse": false,
            "replace": false,
            "merge-extra": false
        }
    }
}
```

The `bin/gpm install` command might need to call `composer update` to ensure all plugin dependencies are updated too.

I'm open to any feedback or suggestions on how to handles plugin dependencies (especially if there are some better alternatives).

**Comment by rhukster** [comment]
This is pretty interesting, but my main concern is that running composer would need to be 100% reliable.  And also we would need to handle uninstalls also.  Not to mention the time that it takes to do composer updates can be quite long and even in some occasions hang. This could potentially lead to your site being left in a broken state.

Would need some real world testing for sure.

**Comment by rhukster** [comment]
I think i'm going to leave this for now.  It really is not a huge issue for a 3rd party plugin to have its own composer based autoloader.  We will reopen and address this at a later date.

**Comment by alexweissman** [comment]
>  It really is not a huge issue for a 3rd party plugin to have its own composer based autoloader.

@rhukster I believe the problem I ran into today is a good counterexample.  My theme was using its own composer-based autoloader to load (of all things) `rockettheme/toolbox`, which is a dependency of my theme, but also of Grav itself.

For whatever reason, trying to load `rockettheme` twice (once by Grav, and once by my theme) led to a strange bug with the cache being improperly managed, and the `page.collections` variable getting set to empty until I cleared my cache manually.  I'm not sure why the problem manifested this way (it took 7 hours to figure out that Composer was the issue!), but when I load my dependencies directly in Grav's main `composer.json` file, the issue no longer occurs.

I will say, that we are using `composer-merge-plugin` [quite effectively](https://learn.userfrosting.com/sprinkles/contents#composerjson) in our framework - I would highly suggest it for Grav as well!
